### PR TITLE
introduce deployment plan

### PR DESCRIPTION
**Pull request by ajssmith**
This changeset introduces the deploymentPlan object to the spec. The plan collects the attributes associated with the deployment enabling edge/interior deployments and a placement attribute for scheduling to the cluster.

### CODE REVIEW

File: deploy/crds/interconnectedcloud_v1alpha1_qdrouterd_cr.yaml [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
-  name: amq-interconnect
+  name: interior-interconnect
 spec:
   # Add fields here
-  count: 3
-  image: quay.io/interconnectedcloud/qdrouterd:1.6.0
-  deploymentMode: lbfrontend
+  deploymentPlan:
+    image: quay.io/interconnectedcloud/qdrouterd:1.6.0
+    role: edge
```
Comment (grs) [inline deploy/crds/interconnectedcloud_v1alpha1_qdrouterd_cr.yaml]: role doesn't match expectation set by name above... is this an interior example or an edge example?

File: deploy/crds/interconnectedcloud_v1alpha1_qdrouterd_crd.yaml [hunk 2]
```diff
@@ -136,6 +169,28 @@ spec:
                   - port
                 type: object
               type: array
+            interEdgeConnectors:
```
Comment (grs) [inline deploy/crds/interconnectedcloud_v1alpha1_qdrouterd_crd.yaml]: Minor issue, but the name doesn't seem quite right to me. Inter-edge sounds like it is between edges. I think I'd prefer just edgeConnectors (and edgeListeners below) though that is not the nicest name either.

File: deploy/crds/interconnectedcloud_v1alpha1_qdrouterd_crd.yaml [hunk 3]
```diff
@@ -136,6 +169,28 @@ spec:
...
+                    format: int32
+                    type: integer
+                  host:
+                    type: string
+                  name:
+                    type: string
+                  port:
+                    format: int32
+                    type: integer
+                  routeContainer:
```
Comment (grs) [inline deploy/crds/interconnectedcloud_v1alpha1_qdrouterd_crd.yaml:188]: Is the routeContainer relevant here? Isn't the underlying role implied by the fact that this is an edge connector?

File: pkg/resources/containers/container.go [hunk 4]
```diff
@@ -129,5 +125,71 @@ func ContainerForQdrouterd(m *v1alpha1.Qdrouterd) corev1.Container {
...
+
+func defaultRequests() corev1.ResourceRequirements {
+	// todo: set based on role
+	rr := corev1.ResourceRequirements{}
+	rr.Requests = make(map[corev1.ResourceName]resource.Quantity)
+	rr.Limits = make(map[corev1.ResourceName]resource.Quantity)
+	rr.Requests[corev1.ResourceCPU] = resource.MustParse("100m")
+	rr.Requests[corev1.ResourceMemory] = resource.MustParse("128Mi")
+	rr.Limits[corev1.ResourceCPU] = resource.MustParse("2000m")
+	rr.Limits[corev1.ResourceMemory] = resource.MustParse("2000Mi")
```
Comment (grs) [inline pkg/resources/containers/container.go]: Should the default be simply not to specify requests/limits?

File: pkg/utils/configs/config.go [hunk 5]
```diff
@@ -43,10 +64,27 @@ func SetQdrouterdDefaults(m *v1alpha1.Qdrouterd) bool {
...
+			m.Spec.InterEdgeListeners = append(m.Spec.InterEdgeListeners, v1alpha1.Listener{
+				Port: 45672,
+				//    Expose: true,
+				//    SslProfile: "default",
+			})
+		}
+	} else if m.Spec.DeploymentPlan.Role == v1alpha1.RouterRoleEdge {
+		if len(m.Spec.InterEdgeConnectors) == 0 {
+			m.Spec.InterEdgeConnectors = append(m.Spec.InterEdgeConnectors, v1alpha1.Connector{
+				Port: 45672,
```
Comment (grs) [inline pkg/utils/configs/config.go:75]: Does it make sense to try and create a default edge uplink connector without knowing the hostname to use?

**Review by grs — APPROVED** [review]
LGTM, minor comments inline.

File: deploy/crds/interconnectedcloud_v1alpha1_qdrouterd_crd.yaml (hunk 3 above)
Comment (ajssmith) [reply deploy/crds/interconnectedcloud_v1alpha1_qdrouterd_crd.yaml:188]: Needed for crd validation, will take action to look into contextualizing the validation...

File: pkg/utils/configs/config.go (hunk 5 above)
Comment (ajssmith) [reply pkg/utils/configs/config.go:75]: totally agree, removed...

File: deploy/crds/interconnectedcloud_v1alpha1_qdrouterd_crd.yaml (hunk 3 above)
Comment (grs) [reply deploy/crds/interconnectedcloud_v1alpha1_qdrouterd_crd.yaml:188]: Of course, I forgot it is in the shared Listener definition. Leave it for now. (Ultimately I think we need to reconsider how we represent route-container connectors/listeners)

**Pull request merged**